core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1233,6 +1233,21 @@ abstract class RDD[T: ClassTag](
           (i, iter) => iter.map((i % curNumPartitions, _))
         }.foldByKey(zeroValue, new HashPartitioner(curNumPartitions))(cleanCombOp).values
       }
+      if (conf.get(ENABLE_EXECUTOR_TREE_AGGREGATE) && partiallyAggregated.partitions.length > 1) {
+        // define a new partitioner that results in only 1 partition
+        val constantPartitioner = new Partitioner {
+          override def numPartitions: Int = 1
+
+          override def getPartition(key: Any): Int = 0
+        }
+        // map the partially aggregated rdd into a key-value rdd
+        // do the computation in the single executor with one partition
+        // get the new RDD[U]
+        partiallyAggregated = partiallyAggregated
```

Comment:
I sort of get the argument about this being more robust on the executor side, but, this does strictly more work - there is always an additional hop from the last agg to the driver now. The amount of work done is no less. I am not convinced this is generally helping, but, it is mostly a guess based on the previous sentence - it's more work

Reply:
We have seen internally in some cases the last fold causes OOM due to the number of values that gets folded. With this approach, the last fold would just be only 1 value.

There is a slight overhead of adding one more hop but overall this should help cases where driver memory is a bottleneck for the last fold operation.

Reply:
As a temporary fix for the internal issues we saw, we increased driver memory and also lowered `spark.maxRemoteBlockSizeFetchToMem` so that it prevents the driver from trying to put the large buffer it is fetching into memory.  Despite the slight overhead, it should prevent the driver from being a single point of failure.

Reply:
Why would this be any less memory intensive after this change? there is still the same last N-to-1 aggregation _somewhere_.

Reply:
it would not be any less memory intensive.  It is just that the executors tend to be configured with higher memory, so there is a lesser chance of this happening on executors. Additionally, executors can retry the task several times, opposed to just 1 try and failing on the driver